protocols/gossipsub/src/error_priv.rs

```diff
@@ -0,0 +1,147 @@
+    TransformFailed(std::io::Error),
+}
+
+impl std::fmt::Display for PublishError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl std::error::Error for PublishError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::SigningError(err) => Some(err),
+            Self::TransformFailed(err) => Some(err),
+            _ => None,
+        }
+    }
+}
+
+/// Error associated with subscribing to a topic.
+#[derive(Debug)]
+pub enum SubscriptionError {
+    /// Couldn't publish our subscription
+    PublishError(PublishError),
+    /// We are not allowed to subscribe to this topic by the subscription filter
+    NotAllowed,
+}
+
+impl std::fmt::Display for SubscriptionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl std::error::Error for SubscriptionError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::PublishError(err) => Some(err),
+            _ => None,
+        }
+    }
+}
+
+impl From<SigningError> for PublishError {
+    fn from(error: SigningError) -> Self {
+        PublishError::SigningError(error)
+    }
+}
+
+/// Errors that can occur in the protocols handler.
+#[derive(Debug, Error)]
+pub enum HandlerError {
+    #[error("The maximum number of inbound substreams created has been exceeded.")]
+    MaxInboundSubstreams,
+    #[error("The maximum number of outbound substreams created has been exceeded.")]
+    MaxOutboundSubstreams,
+    #[error("The message exceeds the maximum transmission size.")]
+    MaxTransmissionSize,
+    #[error("Protocol negotiation timeout.")]
+    NegotiationTimeout,
+    #[error("Protocol negotiation failed.")]
+    NegotiationProtocolError(ProtocolError),
+    #[error("Failed to encode or decode")]
+    Codec(#[from] prost_codec::Error),
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum ValidationError {
+    /// The message has an invalid signature,
+    InvalidSignature,
+    /// The sequence number was empty, expected a value.
+    EmptySequenceNumber,
+    /// The sequence number was the incorrect size
+    InvalidSequenceNumber,
+    /// The PeerId was invalid
+    InvalidPeerId,
+    /// Signature existed when validation has been sent to
+    /// [`crate::behaviour::MessageAuthenticity::Anonymous`].
+    SignaturePresent,
+    /// Sequence number existed when validation has been sent to
+    /// [`crate::behaviour::MessageAuthenticity::Anonymous`].
+    SequenceNumberPresent,
+    /// Message source existed when validation has been sent to
+    /// [`crate::behaviour::MessageAuthenticity::Anonymous`].
+    MessageSourcePresent,
+    /// The data transformation failed.
+    TransformFailed,
+}
+
+impl std::fmt::Display for ValidationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{self:?}")
```

Comment:
We should fix this to be a better text and not just debug. Can be in a follow-up though!

Reply:
I've put something more, but not sure if that's what you mean :thinking:

Reply:
Not quite. The debug print of the error will just render its name. The doc strings on the variants itself are actually quite good.

I'd prefer it in a separate PR though as it is unrelated to the module being private now.

Would you mind reverting that commit? Everything else looks 💯

Reply:
done :heavy_check_mark:

Reply:
If you have a look at the doc strings on the variants, those would provide a much better error message than their debug print which will just print the variants name.

You can write a test where you `to_string` a variant to explore the difference :)